### ISSUE TITLE
Support for Other Modules

### ISSUE THREAD

**Issue by allenkim** [facebookresearch/CrypTen#137]
Hi,

I was wondering about support for more modules such as max, min, or random tensors while constructing a neural network. Would these features be easily extendable or is there some inherent limitation? I would assume since ReLU and Dropout are supported, that functions like the ones mentioned should be possible, but I wasn't sure what the best approach would be.

Thank you!

**Comment by marksibrahim** [comment]
Hi Allenkim, 

Is there a particular reason you're looking for `max` or `min` as modules? 

Similar to PyTorch you can call functions such as `relu()` or `max()` in the forward function of your model. For example,

```
import crypten 

class SampleNetwork(crypten.nn.Module):
    def __init__(self, n_features=N_FEATURES):
        super().__init__()
        self.linear = crypten.nn.Linear(n_features, 10)

    def forward(self, x):
        return self.linear(x).max()
```

Let me know if you still have questions.

**Comment by allenkim** [comment]
I was looking to experiment with some custom activation functions instead of ReLU or sigmoid, but I needed functions such as min/max(x,y) and the ability to generate randomness for what I wanted to try.

**Comment by marksibrahim** [comment]
Could you provide an example of what you're hoping to accomplish?

Similar to PyTorch we don't implement `max` or `min` as `nn.Modules`, but we do provide `max` and `min` functions on CrypTensors, see https://crypten.readthedocs.io/en/latest/cryptensor.html#crypten.cryptensor.CrypTensor.max

**Comment by allenkim** [comment]
As an example, let's say we are using ReLU as our activation function. Before running our values through ReLU, I want to be able to add a random noisy tensor before doing so, but I wasn't sure if there was support for generating random tensors. Regarding max and min, I wanted to be able to clip the values between 0 and 1 for example, but in theory, this can be emulated with ReLU.

**Comment by marksibrahim** [comment]
Thanks, I see. You have a couple of options: 

1) We provide random noise from uniform and Bernoulli distributions. For example `crypten.rand(2, 3)` or `crypten.bernoulli(torch.tensor(0.10))` 
2) You can generate a random CrypTensor from a random PyTorch tensor. If you're running in a multi-party setting, then you should broadcast the shares to the other parties. Checkout https://github.com/facebookresearch/CrypTen/blob/bc0ab844d9f5648a87b0b94bdd05b97f890b4b94/test/multiprocess_test_case.py#L21 for an example of how we create random tensors for our tests as well as https://github.com/facebookresearch/CrypTen/blob/bc0ab844d9f5648a87b0b94bdd05b97f890b4b94/tutorials/Tutorial_1_Basics_of_CrypTen_Tensors.ipynb for how to create CrypTensors from PyTorch tensors.

**Comment by knottb** [comment]
`crypten.bernoulli()` can also be called using a CrypTensor `x` as its input distribution (i.e. the distribution is parameterized by `x`) using `x.bernoulli()`. (It may also be possible by calling `crypten.bernoulli(x)`.)

**Comment by allenkim** [comment]
Thanks for the information! I see that CrypTensors support all the operations I need. Maybe this is a silly question but looking further, I see that in examples using neural networks, one defines an example network using standard PyTorch syntax and then constructs the encrypted model. 

`model_plaintext = ExampleNet()`
`dummy_input = torch.empty(1, 1, x, x)`
`model = crypten.nn.from_pytorch(model_plaintext, dummy_input)`
`model.encrypt()`

I'm assuming when we define the model, we use standard PyTorch syntax and at this point, CrypTen operations aren't capable of being mixed in. If I want to make the changes I mentioned before, is there a way to define it in standard PyTorch syntax and have it converted or would I have to define the entire process with CrypTen? 

Thanks for the help!

**Comment by marksibrahim** [comment]
I'm glad that was helpful. Yes, you can import a network defined in PyTorch as you describe using `model = crypten.nn.from_pytorch(model_plaintext, dummy_input)`. 

Note however, we don't support importing all PyTorch operations just yet. We do support many common architectures (see [Tutorial 7](https://github.com/facebookresearch/CrypTen/blob/master/tutorials/Tutorial_7_Training_an_Encrypted_Neural_Network.ipynb) for examples). You can also always define your network in CrypTen. The syntax and structure should be very similar to that of PyTorch.

**Comment by allenkim** [comment]
Thanks. I saw those examples, and originally raised this issue regarding if there was support for lesser common features in common architectures.

Also, regarding defining the network in CrypTen, if I understand, this is without any ``nn`` features. Is there a reference one can look at to do so? Or is it just a matter of following the same PyTorch approach, but using CrypTensors instead of regular Tensors. I imagine it may be similar, but there would probably still be some CrypTen specific things that one would have to do.

**Comment by marksibrahim** [comment]
The `crypten.nn` section of the documentation is a good reference (https://crypten.readthedocs.io/en/latest/nn.html). This contains extensive documentation of supported CrypTen modules and examples of the syntax for defining both `nn.Module` and `nn.Sequential` CrypTen networks.

**Comment by marksibrahim** [comment]
Feel free to re-open if you still have questions or create a new issue if you encounter one.
